### PR TITLE
1150 add pattern matching

### PR DESCRIPTION
**Pull request by OlesiaSub**
#1150
I added pattern matching, added a test and also tested it on fibonacci. It works fine with `org.eolang.error` and with `org.*.error`. But there is a problem with `org.*`: building of fibonacci takes a really long time (I waited for ~10 minutes and it didn't finish) and I don't know why (maybe it's an expected behaviour...)

### CODE REVIEW

**Comment by OlesiaSub** [comment]
Not all checks pass, afaik connected to #1178

**Comment by OlesiaSub** [comment]
@yegor256 pls check

**Comment by yegor256** [comment]
@OlesiaSub I can't understand why we need `gmiIncludes` and `includeMatchers` at the same time. I was thinking that we will set `gmiIncludes` to `["org.*", "org.eolang.io.*", "foo.a", "foo.**.b"]`. Here `foo.**.b` means `foo.x.y.z.b` and `foo.b`.

In other words, `**` means anything including `dot`, while `*` means anything except the dot.

**Comment by OlesiaSub** [comment]
@yegor256 do you want me to implement my own pattern matcher? Afaik standard regexes in java don't support ** syntax. I think using standard regexes is a more universal solution which is commonly used everywhere.
Smth like this will work: `"org.[^.]*.error"` matches only expressions with one string between the dots. `"org.*.error"` matches expressions with anything between the dots.

Btw, foo.**.b can't match foo.b because you explicitly set two dots in the first expression and the second one has only one dot.

**Comment by OlesiaSub** [comment]
@yegor256 answering your first question, I need to convert Strings from gmiIncludes to Patterns. Alternatively I could map Strings to Patterns for every check, but it would be more time consuming.

**Comment by yegor256** [comment]
@OlesiaSub maybe we can make a new open source library, doing glob-ish matching of two strings?)

**Comment by OlesiaSub** [comment]
@yegor256 ok got it. Can I just use this then?
```
val a = FileSystems.getDefault().getPathMatcher("glob:org.**.error")
val res = a.matches(Path.of("org.eolang.smth.error"))
```
`res` will be true in this case.

**Comment by yegor256** [comment]
@OlesiaSub I don't think it works correctly with dots. Try this:

```
val a = FileSystems.getDefault().getPathMatcher("glob:org.*.error")
val res = a.matches(Path.of("org.eolang.smth.error"))
```

The result will be `true`, but should be `false`.

**Comment by OlesiaSub** [comment]
@yegor256 ok I see. Then can I just support * and **? or does my matcher have to support something else?

But still please consider using regular expressions, I'll explain my viewpoint. I think it's a more common way to do pattern matching. Plus, in the future people might want to add complex patterns, which won't be covered by * and ** syntax. So either we'll have to implement a huge parser to cover all the possible cases or we may just stick with an existing mechanism of regular expressions.

**Comment by yegor256** [comment]
@OlesiaSub I think it should be a very simple patcher, which (as you said) supports only `*` and `**`. I would even implement it using regular expressions. First, you convert all `*` to `[a-z0-9]+` and all `**` to `[a-z0-9.]+?`, and then just use Java `Pattern`. Should be pretty easy.

**Comment by OlesiaSub** [comment]
@yegor256 ok sounds good, thanks) I probably misunderstood you and thought that you wanted to completely eliminate usage of regexes, which seemed strange to me

**Comment by OlesiaSub** [comment]
@yegor256 pls check

**Comment by yegor256** [comment]
@OlesiaSub again, why do we need _four_ parameters of this Maven goal? Just two will be enough: `gmiIncludes` and `gmiExcludes`. This is how I want to configure this plugin:

```
      <plugin>
        <groupId>org.eolang</groupId>
        <artifactId>eo-maven-plugin</artifactId>
        <executions>
          <execution>
            <goals>
              <goal>gmi</goal>
            </goals>
            <configuration>
              <gmiIncludes>
                <o>org.eolang.**</o>
                <o>org.example.abc</o>
              </gmiIncludes>
            </configuration>
          </execution>
      </executions>
```

That's all.

**Comment by OlesiaSub** [comment]
@yegor256 agreed, done

File: eo-maven-plugin/src/main/java/org/eolang/maven/GmiMojo.java [hunk 1]
```diff
@@ -238,9 +236,15 @@ public void exec() throws IOException {
         final Path home = this.targetDir.toPath().resolve(GmiMojo.DIR);
         int total = 0;
         int instructions = 0;
+        final Set<Pattern> includes = this.gmiIncludes.stream()
+            .map(i -> Pattern.compile(this.createMatcher(i)))
+            .collect(Collectors.toSet());
+        final Set<Pattern> excludes = this.gmiExcludes.stream()
+            .map(Pattern::compile)
```
Comment (yegor256) [inline eo-maven-plugin/src/main/java/org/eolang/maven/GmiMojo.java:243]: @OlesiaSub why not `createMatcher` here too?

File: eo-maven-plugin/src/main/java/org/eolang/maven/GmiMojo.java [hunk 2]
```diff
@@ -203,16 +205,12 @@ public final class GmiMojo extends SafeMojo {
...
      *  properties since there is no way of passing it via command line.
      * @checkstyle MemberNameCheck (15 lines)
-     * @todo #1146:30min At the moment we don't support pattern matching, but
-     *  double-star means "everything". Let's implement proper matching,
-     *  where "org.eolang.int" would be matched by "org.*.int" and by
-     *  "org.**". The same is true about gmiExclude, let's fix it too.
      */
     @Parameter
-    private Set<String> gmiIncludes = new SetOf<>("**");
+    private Set<String> gmiIncludes = new SetOf<>("[A-Za-z0-9.]+?");
```
Comment (yegor256) [inline eo-maven-plugin/src/main/java/org/eolang/maven/GmiMojo.java:210]: @OlesiaSub `**` would be more logical to put here, I believe

File: eo-maven-plugin/src/test/java/org/eolang/maven/GmiMojoTest.java [hunk 3]
```diff
@@ -67,7 +68,7 @@ void bigSlowTest() throws Exception {
             }
             program.append("[x y z] > foo\n");
         }
-        final XML graph = GmiMojoTest.toGraph(program.toString());
+        final XML graph = GmiMojoTest.toGraph(program.toString(), null);
```
Comment (yegor256) [inline eo-maven-plugin/src/test/java/org/eolang/maven/GmiMojoTest.java:71]: @OlesiaSub this is a terrible practice: passing as an argument

Comment (yegor256) [reply eo-maven-plugin/src/test/java/org/eolang/maven/GmiMojoTest.java:71]: @OlesiaSub why not `**`?

File: eo-maven-plugin/src/main/java/org/eolang/maven/GmiMojo.java (hunk 2 above)
Comment (OlesiaSub) [reply eo-maven-plugin/src/main/java/org/eolang/maven/GmiMojo.java:210]: In my implementation ** matches only expressions containing a dot. Beacuse I though you didn't want org.**.foo to match expressions like org.bar.foo, where it's only one word between the dots. Or did I misunderstand you and it should match anything?

**Comment by yegor256** [comment]
@OlesiaSub see above some comments

File: eo-maven-plugin/src/test/java/org/eolang/maven/GmiMojoTest.java (hunk 3 above)
Comment (OlesiaSub) [reply eo-maven-plugin/src/test/java/org/eolang/maven/GmiMojoTest.java:71]: @yegor256 fixed, I didn't really pay attention to that test as it's disabled, so I forgot to fix it afterwards

File: eo-maven-plugin/src/main/java/org/eolang/maven/GmiMojo.java (hunk 2 above)
Comment (OlesiaSub) [reply eo-maven-plugin/src/main/java/org/eolang/maven/GmiMojo.java:210]: @yegor256 see above

Comment (yegor256) [reply eo-maven-plugin/src/main/java/org/eolang/maven/GmiMojo.java:210]: @OlesiaSub `**` means anything _including_ the dot, while `*` means anything _excluding_ the dot

**Comment by OlesiaSub** [comment]
@yegor256 fixes comments

**Comment by yegor256** [comment]
@rultor merge

**Pull request merged**